src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DocumentImportsTracker.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
+{
+    internal class DocumentImportsTracker
+    {
+        private readonly object _lock;
+
+        private IReadOnlyList<DocumentSnapshot> _imports;
+
+        public DocumentImportsTracker()
+        {
+            _lock = new object();
+        }
+
+        public IReadOnlyList<DocumentSnapshot> GetImports(ProjectSnapshot project, DocumentSnapshot document)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (_imports == null)
+            {
+                lock (_lock)
+                {
+                    if (_imports == null)
+                    {
+                        _imports = GetImportsCore(project, document);
+                    }
+                }
+            }
+
+            return _imports;
+        }
+
+        private IReadOnlyList<DocumentSnapshot> GetImportsCore(ProjectSnapshot project, DocumentSnapshot document)
+        {
+            var projectEngine = project.GetProjectEngine();
+            var importFeature = projectEngine.ProjectFeatures.OfType<IImportProjectFeature>().FirstOrDefault();
+            var projectItem = projectEngine.FileSystem.GetItem(document.FilePath);
+            var importItems = importFeature?.GetImports(projectItem).Where(i => i.Exists);
+            var imports = new List<DocumentSnapshot>();
+            foreach (var item in importItems)
+            {
+                if (item.PhysicalPath == null)
+                {
+                    // This is a default import.
+                    var defaultImport = new DefaultImportDocumentSnapshot(project, item);
```

Comment:
I still really don't get why this has to be backed by a snapshot. Why isn't this something more primitive like a `SourceText`.

Reply:
Because anything outside the `Tracker` wouldn't have the knowledge of what an import is. I don't like the idea of making default imports first class on `DocumentSnapshot`. The only other way without doing that would be to introduce an `ImportSnapshot` type which can provide the SourceText and Version of the import and would be backed by a `DocumentSnapshot` for real imports and just text for default imports. Do you like this better?